### ISSUE TITLE
Implement Rename

### ISSUE THREAD

**Issue by shvachko** [GiraffaFS/giraffa#29]
Original [issue 29](https://code.google.com/p/giraffa/issues/detail?id=29) created by shvachko on 2012-09-14T17:35:00.000Z:

We need a way of renaming files in the namespace. This will be difficult -- and even harder to do atomically. We should devise a strategy to deal with this issue.

My vision so far:

In-place rename is relatively straightforward. We need to grab the file row using the RowKey, copy its info into a new row, with a new RowKey, and insert it back into HBase, and delete the old row using the old RowKey; we then need to update the parent directory's DirectoryTable to reflect the RowKey change.

A file &quot;move&quot; rename is a little harder... here is my vision: We need to do the same as the in-place mostly, except we need to add the RowKey to the &quot;new&quot; parent directory's DirectoryTable, and just delete it from the previous parent directory's DirectoryTable.

The difficulty comes in dealing with directories. Renaming a directory, even in-place, is ALOT harder... because now all the RowKeys under that directory, and the old and new parent directory's DirectoryTables need to be updated to reflect the change. This could potentially take a lot of time as we would be constantly deleting and creating new RowKeys, and making HBase calls to update information everywhere.

So renaming files seems easy, but renaming directories is a potential major issue.

**Comment by shvachko** [comment]
Comment [#1](https://code.google.com/p/giraffa/issues/detail?id=29#c1) originally posted by shvachko on 2012-09-16T07:16:19.000Z:

Here is a quick work of what I am proposing.

**Comment by shvachko** [comment]
Comment [#2](https://code.google.com/p/giraffa/issues/detail?id=29#c2) originally posted by shvachko on 2012-09-16T20:44:39.000Z:

Here is further work / optimization I have done so far.

**Comment by shvachko** [comment]
Comment [#3](https://code.google.com/p/giraffa/issues/detail?id=29#c3) originally posted by shvachko on 2012-09-17T15:44:33.000Z:

It looks like the only options for rename are NONE and OVERWRITE; where OVERWRITE will overwrite the destination file if it already exists. This parsing should be easy to add in.

In HDFS, if you rename a directory to a directory that already exists, and you "overwrite", will the contents of the old directory go away? Or do they get merged into?

**Comment by shvachko** [comment]
Comment [#4](https://code.google.com/p/giraffa/issues/detail?id=29#c4) originally posted by shvachko on 2012-09-17T18:28:40.000Z:

Well, in Giraffa according to our design we will support "in-place" atomic renames. That is we will allow to change local name of a file, but will not allow moving it into a different directory.
But in order to implement the in-place we will need to introduce inode numbers and implement new RowKey, which uses indoe numbers as file keys.
In your patch you can see that the rename is not atomic. Imagine two threads trying to move the same file. In HDFS it is happening on a single NameNode, and one can use locks to ensure exclusivity of the move. In distributed setting the same will not work.

**Comment by shvachko** [comment]
Comment [#5](https://code.google.com/p/giraffa/issues/detail?id=29#c5) originally posted by shvachko on 2012-09-21T05:25:02.000Z:

There are many more components on this that block this task from being complete. I am moving this down to low priority for now and will assign tasks that block this from completion with higher priority.

**Comment by shvachko** [comment]
Comment [#6](https://code.google.com/p/giraffa/issues/detail?id=29#c6) originally posted by shvachko on 2012-09-21T05:33:45.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#7](https://code.google.com/p/giraffa/issues/detail?id=29#c7) originally posted by shvachko on 2012-09-25T07:20:31.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#8](https://code.google.com/p/giraffa/issues/detail?id=29#c8) originally posted by shvachko on 2012-09-25T07:53:58.000Z:

&lt;empty&gt;

**Comment by shvachko** [comment]
Comment [#9](https://code.google.com/p/giraffa/issues/detail?id=29#c9) originally posted by shvachko on 2012-09-28T19:07:22.000Z:

https://issues.apache.org/jira/browse/HBASE-3584

I think with HBase 0.94.0 we can offer in-place atomic rename with this JIRA. It is probably not worth pursing but maybe for the sake of getting the "mv" commands to pass for issue 15 we should consider it.

Actually now that I think about it, we should provide some sort interface mechanism for rename based on the RowKey we are using -- so that if we choose to use FullPath, or INodeNumber, or RandomNumber, we should have different rename implementations for each RowKey... just a though. Let me know what you think.

This is just thought at the moment -- no patches or work to be done just yet; just want your opinion.

**Comment by shvachko** [comment]
Comment [#10](https://code.google.com/p/giraffa/issues/detail?id=29#c10) originally posted by shvachko on 2012-12-08T19:22:46.000Z:

Now that we have eliminated DirectoryTables, file renames become a matter of creating a new row and deleting the old one; directory renames are still much harder because everything below that directory needs to be re-created (if we are still using FPRK).
These issues will be solved by INodeRowKey.

**Comment by shvachko** [comment]
Comment [#11](https://code.google.com/p/giraffa/issues/detail?id=29#c11) originally posted by shvachko on 2013-08-20T17:03:07.000Z:

Milan, we should post up your rename algorithm here. Also, as an amendment, we can actually attempt to take out a rowLock on the destination row as well (which would only be successful if they ARE on the same RegionServer).

**Comment by shvachko** [comment]
Comment [#12](https://code.google.com/p/giraffa/issues/detail?id=29#c12) originally posted by shvachko on 2013-08-20T18:43:58.000Z:

The algorithm for in-place file rename (without overwrite handling) is as follows:

The basic idea is that to rename "fileA" to "fileB", we PUT a new row for "fileB" with the same exact cells as "fileA", and then DELETE the row for "fileA". We place a row lock on the source row to ensure that the filesystem is blocked from doing anything to the files while they are renaming. The additional complexity in the algorithm is for handling recovery procedures, which kick in when the server dies in the middle of this non-atomic operation. The goals for recovery are (1) when rename fails because the server dies and is called again, the method should pick up where it left off; (2) when a server sees temporary rename files idling, it reverts the rename or continues the rename, whichever is feasible. The bottom line is that the user should not lose data and the system should not become cluttered with temp files because the rename failed.

Every iNode has an additional column for storing a rename status object called RENAME_IN_PROGRESS. The object has two fields: "flag" and "src". On any file, when "flag" is 'true', this is an indication that the file is currently being renamed from a source file whose row key is "src". It is used to handle recovery procedures. A file whose RENAME_IN_PROGRESS flag is set is considered a temporary file and is hidden and locked from the user. In this algorithm, such a file is represented by appending [R] to the file name; i.e. dst[R].

Procedure for renaming "src" to "dst"

1: LOCK src
2: RECOVER (jump ahead to appropriate step)
3: PUT dst[R]
4: UNLOCK+DELETE src (technically something could happen between the unlock and delete; is that something to consider?)
5: PUT dst (remove the [R] flag)

Recovery Procedure:

Case 1: src exists, dst[R] exists: GO TO (4)
Case 2: src does not exist, dst[R] exists: GO TO (5)
Otherwise: rename has not started; check for overwrite or file not found errors and GO TO (3)

Additionally, we will have a daemon that searches for idle [R] files and either continues the rename or reverts it. The details for this need to be worked out.

We also need to handle what happens when a lock is taken out on a row and the server dies before we unlock. What happens to the lock?

**Comment by shvachko** [comment]
Comment [#13](https://code.google.com/p/giraffa/issues/detail?id=29#c13) originally posted by shvachko on 2013-08-20T18:49:07.000Z:

Plamen, I think taking a row lock on the destination is unnecessary, because it only works when the src and dst are on the same region server. In this case, we can already do an atomic PUT+DELETE since the api provides for atomic multi-row mutations on the same region server, eliminating the need for locks entirely. (for anyone else who's reading, this is an optimization not in the algorithm)

**Comment by shvachko** [comment]
Comment [#14](https://code.google.com/p/giraffa/issues/detail?id=29#c14) originally posted by shvachko on 2013-08-20T19:02:20.000Z:

Attached is a draft patch for file in-place rename (no overwrites) with test cases. The test cases fail because they are unable to test rename cases that fail. The method rename2() returns a void, and the only way to indicate failure is through exceptions, which are not currently passed through the hbase rpc. So we need to get Issue75 working before we can get the tests to pass.

**Comment by shvachko** [comment]
Comment [#15](https://code.google.com/p/giraffa/issues/detail?id=29#c15) originally posted by shvachko on 2013-08-20T19:03:55.000Z:

The patch is large because a new proto was created for the RENAME_IN_PROGRESS status object and this and some other protos were moved into GiraffaProtos

**Comment by shvachko** [comment]
Comment [#16](https://code.google.com/p/giraffa/issues/detail?id=29#c16) originally posted by shvachko on 2013-08-21T00:45:57.000Z:

Good work Milan. No worries about the size; this is a big feature now so large size is acceptable.

Here are my comments:

1a) Let's rename RENAME_IN_PROGRESS to RENAME_STATE (and bytes to "renameState")
1b) Let's rename STATE to FILE_STATE (and bytes to "fileState")
2) Make GiraffaFileSystem do: return grfaClient.rename(), not true all the time.
3) Let's rename RenameInProgressProto to RenameStateProto.
4) Why not just attempt to do getINode first? That way you can get the RowKey right away rather than generate the RowKey twice.
5) Can we call delete with the rowLock? Used to be possible, not sure if it is anymore. If we cannot, is it worthwhile to use rowLocks anymore?
6) Would be good to test recovery as well in unit test.

Sorry for the renames! Good work so far!

**Comment by shvachko** [comment]
Comment [#17](https://code.google.com/p/giraffa/issues/detail?id=29#c17) originally posted by shvachko on 2013-08-21T03:35:25.000Z:

Thanks!

2) GiraffaFileSystem.rename(src,dst) calls grfaClient.rename(src,dst,options), which calls ClientProtocol.rename2, which returns void (errors are handled through exceptions). To return a boolean, we would have to call the deprecated grfaClient.rename(src,dst), which calls ClientProtocal.rename so it returns a boolean. Alternatively we could catch non-IO exceptions and return false on those

5) We cannot call delete with the row lock, so we have to unlock and then delete. I don't know if it's still worthwhile. The lock prevents the user from doing anything to the source file until we have created the destination file. The problem arises when the user replaces the source file in-between the unlock and delete. Then our rename will delete that user's replacement file. It seems like we may have to implement our own INode lock, which would render the row locks unnecessary.

6) I'm not sure how to test recovery using unit tests. I think you mentioned a possible way last week, but I don't remember.

**Comment by shvachko** [comment]
Comment [#18](https://code.google.com/p/giraffa/issues/detail?id=29#c18) originally posted by shvachko on 2013-08-21T22:52:38.000Z:

Milan,

As a response to 6), I'm attaching an example of what to do for testing recovery -- look at the "InplaceRenameFromRecoveryStage2" test.

As for 2), be careful, there are some places where you are relying on GiraffaFS.rename() to return false, and it will ALWAYS return true currently.

5) Agreed -- I've looked. Couldn't find a way; its okay.

**Comment by shvachko** [comment]
Comment [#19](https://code.google.com/p/giraffa/issues/detail?id=29#c19) originally posted by shvachko on 2014-05-04T05:26:27.000Z:

Hey Milan; this work has gotten stale. Is it possible to get a fresh patch? I understand its been a long time but let me know.

**Comment by shvachko** [comment]
Comment [#20](https://code.google.com/p/giraffa/issues/detail?id=29#c20) originally posted by shvachko on 2014-05-05T18:10:18.000Z:

Hey Plamen - I get done with my finals the end of this week, and then i'll look into this again this weekend

**Comment by shvachko** [comment]
Comment [#21](https://code.google.com/p/giraffa/issues/detail?id=29#c21) originally posted by shvachko on 2014-05-24T07:50:52.000Z:

The attached patch makes the changes commented on my previous patch. Recovery testing was added and successfully passes for file renames/moves. Directory renames are not yet implemented. Overwrites are disallowed. Row Locks were removed as per the previous discussion in the commentary; we need to implement our own INode locks to get desired functionality. Once we implement INodeIDRowKey, we can significantly optimize for in-place renames of files and directories. We can also optimize for file renames that stay within the same region server by using atomic put+delete.

INode Locks, INodeIDRowKey, and Exception Handling are unresolved issues crucial for rename to work optimally.

**Comment by shvachko** [comment]
Comment [#22](https://code.google.com/p/giraffa/issues/detail?id=29#c22) originally posted by shvachko on 2014-06-02T15:30:52.000Z:

1) Should recover state become an enum? Would be easier to understand what each state integer means.
2) There are unused methods in TestRename; are saving them for Dir rename implementation?
3) Perhaps for future issue, but should we also move the 'getINodeField' static methods from NamespaceProcessor into something like INodeUtils? Just a though -- doesn't need to be done as part of this issue.

Other than that, great stuff here! I like the recovery implementation, pretty straightforward now in the code. And rename itself is coming together; I hope its enough to kick off some MapReduce jobs. :)

Let's handle these suggestions and then commit.

**Comment by shvachko** [comment]
Comment [#23](https://code.google.com/p/giraffa/issues/detail?id=29#c23) originally posted by shvachko on 2014-06-03T18:20:58.000Z:

Hey Plamen, I made an enum for RenameRecoveryState in NamespaceProcessor directly after rename2(). The unused methods are indeed for a future dir rename implementation. I added @SupressWarnings for now. I also simplified RenameTest and made it more unit-test style. The new patch is built off the patch for Issue 75 (Exception handling) since it's about to get committed.

Moving the getINodeField static methods to INodeUtils definitely makes more sense for a future issue; they are out of place right now in NamespaceProcessor.

Also, any ideas why my patches don't have the "view" option on Google Code?

**Comment by shvachko** [comment]
Comment [#24](https://code.google.com/p/giraffa/issues/detail?id=29#c24) originally posted by shvachko on 2014-06-06T05:13:51.000Z:

I don't have a "view" option either. I am not sure if there is a way to enable it. Konstantin is the admin so only he would be able to.

**Comment by shvachko** [comment]
Comment [#25](https://code.google.com/p/giraffa/issues/detail?id=29#c25) originally posted by shvachko on 2014-06-06T05:32:01.000Z:

TestGiraffaCLI shows 94% passing tests.
Only 33 failing tests overall now: move directory and the dfsadmin commands!

I got the following exceptions however in TestRename:
{{{
Running org.apache.giraffa.TestRename
Tests run: 10, Failures: 0, Errors: 4, Skipped: 0, Time elapsed: 51.011 sec <<< FAILURE!
testFileInPlaceOverwrite(org.apache.giraffa.TestRename)  Time elapsed: 4.804 sec  <<< ERROR!
java.lang.Exception: Unexpected exception, expected<org.apache.hadoop.fs.FileAlreadyExistsException> but was<java.lang.AssertionError>
    at org.junit.Assert.fail(Assert.java:91)
    at org.junit.Assert.assertTrue(Assert.java:43)
    at org.junit.Assert.assertFalse(Assert.java:68)
    at org.junit.Assert.assertFalse(Assert.java:79)
    at org.apache.giraffa.TestRename.rename(TestRename.java:140)
    at org.apache.giraffa.TestRename.testFileInPlaceOverwrite(TestRename.java:217)

testFileInPlaceDestinationIsDirectory(org.apache.giraffa.TestRename)  Time elapsed: 4.734 sec  <<< ERROR!
java.lang.Exception: Unexpected exception, expected<java.io.IOException> but was<java.lang.AssertionError>
    at org.junit.Assert.fail(Assert.java:91)
    at org.junit.Assert.assertTrue(Assert.java:43)
    at org.junit.Assert.assertFalse(Assert.java:68)
    at org.junit.Assert.assertFalse(Assert.java:79)
    at org.apache.giraffa.TestRename.rename(TestRename.java:140)
    at org.apache.giraffa.TestRename.testFileInPlaceDestinationIsDirectory(TestRename.java:223)

testFileMoveInvalidDestination(org.apache.giraffa.TestRename)  Time elapsed: 4.734 sec  <<< ERROR!
java.lang.Exception: Unexpected exception, expected<java.io.IOException> but was<java.lang.AssertionError>
    at org.junit.Assert.fail(Assert.java:91)
    at org.junit.Assert.assertTrue(Assert.java:43)
    at org.junit.Assert.assertTrue(Assert.java:54)
    at org.apache.giraffa.TestRename.rename(TestRename.java:139)
    at org.apache.giraffa.TestRename.testFileMoveInvalidDestination(TestRename.java:247)

testFileMoveOverwrite(org.apache.giraffa.TestRename)  Time elapsed: 4.783 sec  <<< ERROR!
java.lang.Exception: Unexpected exception, expected<org.apache.hadoop.fs.FileAlreadyExistsException> but was<java.lang.AssertionError>
    at org.junit.Assert.fail(Assert.java:91)
    at org.junit.Assert.assertTrue(Assert.java:43)
    at org.junit.Assert.assertFalse(Assert.java:68)
    at org.junit.Assert.assertFalse(Assert.java:79)
    at org.apache.giraffa.TestRename.rename(TestRename.java:140)
    at org.apache.giraffa.TestRename.testFileMoveOverwrite(TestRename.java:254)
}}}

Seems somehow renames with overwrites are failing. I think it's because in the test itself you aren't actually specifying it with the overwrite flag. Look at TestRename line 138.

**Comment by shvachko** [comment]
Comment [#26](https://code.google.com/p/giraffa/issues/detail?id=29#c26) originally posted by shvachko on 2014-06-08T21:07:25.000Z:

Looks like issue#&nbsp;75 fixes the rename test.
Let's commit both if Plamen has a go on this.

**Comment by shvachko** [comment]
Comment [#27](https://code.google.com/p/giraffa/issues/detail?id=29#c27) originally posted by shvachko on 2014-06-09T06:06:29.000Z:

Yeah the patch was built off of Issue 75 (i.e. it throws and catches Exceptions) since that was about to be committed. Those tests were failing because it was probably run on the pre-Issue75 trunk which can't handle Exceptions. Also, I don't specify the overwrite flag because overwrites aren't supported yet and the flag is ignored.

**Comment by shvachko** [comment]
Comment [#28](https://code.google.com/p/giraffa/issues/detail?id=29#c28) originally posted by shvachko on 2014-06-12T18:42:26.000Z:

Looks good Milan! +1 from me. Please commit.

Unrelated, but, even without your patch, I am getting the parent pom.xml unable to run "mvn clean test -DskipTests".

Something about path not being a source dir?
{{{
[INFO] Configured Artifact: org.apache.hadoop:giraffa-core:conf:0.0.2-SNAPSHOT:zip
[INFO] Unpacking /Users/plamenjeliazkov/giraffa/giraffa-core/target/classes to
  /Users/plamenjeliazkov/giraffa/giraffa-standalone/target/giraffa-standalone/giraffa/conf
   with includes null and excludes:null
org.codehaus.plexus.archiver.ArchiverException: The source must not be a directory.
    at org.codehaus.plexus.archiver.AbstractUnArchiver.validate(AbstractUnArchiver.java:174)
    at org.codehaus.plexus.archiver.AbstractUnArchiver.extract(AbstractUnArchiver.java:107)
}}}

**Comment by shvachko** [comment]
Comment [#29](https://code.google.com/p/giraffa/issues/detail?id=29#c29) originally posted by shvachko on 2014-06-13T06:50:29.000Z:

Plamen this is known issue. See
https://jira.codehaus.org/browse/MDEP-98
This is because we have that conf.zip file. If we must zip it, then there is one way to avoid the error to not call compile or dependent targets like test.
If you run "mvn clean package" it succeeds.

My +1 to commit.

**Comment by shvachko** [comment]
Comment [#30](https://code.google.com/p/giraffa/issues/detail?id=29#c30) originally posted by shvachko on 2014-06-13T19:24:18.000Z:

Committed
